jbpm-services/jbpm-services-ejb/jbpm-services-ejb-timer/src/main/java/org/jbpm/services/ejb/timer/EJBTimerScheduler.java

```diff
@@ -59,46 +78,116 @@
-				e.printStackTrace();
-			}
+                Thread.sleep(500);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
             time += 500;
 
             if (time > OVERDUE_WAIT_TIME) {
-            	logger.debug("No timer service found after waiting {} ms", time);
-            	break;
+                logger.debug("No timer service found after waiting {} ms", time);
+                break;
             }
         }
-		try {
-			((Callable<Void>) timerJobInstance).call();
-		} catch (Exception e) {
-			logger.warn("Execution of time failed due to {}", e.getMessage(), e);
-			throw new RuntimeException(e);
+        try {
+            transaction(this::executeTimerJobInstance, timerJobInstance);
+        } catch (Exception e) {
+            recoverTimerJobInstance(timerJob, e);
 		}
 	}
 
+    private void executeTimerJobInstance(TimerJobInstance timerJobInstance) throws Exception {
+        try {
+            ((Callable<?>) timerJobInstance).call();
+        } catch (Exception e) {
+            logger.warn("Execution of time failed due to {}", e.getMessage(), e);
+            throw e;
+        }
+    }
+
+    private void recoverTimerJobInstance(EjbTimerJob ejbTimerJob, Exception e) {
+        switch (TIMER_RETRY_POLICY) {
+            case RETRY:
+                logger.warn("Execution of time failed. The timer will be retried {}", ejbTimerJob.getTimerJobInstance());
+                Transaction<TimerJobInstance> operation = (instance) -> {
+                    ZonedDateTime nextRetry = ZonedDateTime.now().plus(TIMER_RETRY_INTERVAL, ChronoUnit.MILLIS);
+                    EjbTimerJobRetry info = null;
+                    if(ejbTimerJob instanceof EjbTimerJobRetry) {
+                        info = ((EjbTimerJobRetry) ejbTimerJob).next();
+                    } else {
+                        info =  new EjbTimerJobRetry(instance);
+                    }
+                    if (info.getRetry() > TIMER_RETRY_LIMIT) {
+                        logger.warn("The timer {} reached retry limit {}. It won't be retried again", instance, TIMER_RETRY_LIMIT);
+                    }
+                    TimerConfig config = new TimerConfig(info, true);
+                    timerService.createSingleActionTimer(Date.from(nextRetry.toInstant()), config);
+                };
+                try {
+                    transaction(operation, ejbTimerJob.getTimerJobInstance());
+                } catch (Exception e1) {
+                    logger.error("Failed to executed timer recovery {}", e1.getMessage(), e1);
+                }
+                break;
+            case IGNORE:
+                logger.warn("Execution of time failed. The timer won't be retried {}", ejbTimerJob.getTimerJobInstance());
+                break;
+            case PLATFORM:
+                logger.warn("Execution of time failed. Application server policy applied {}", ejbTimerJob.getTimerJobInstance());
```

Comment:
does this differ between app servers? we use WAS and WLS with EJB timer service as far as I know.

Reply:
it might be... I don't know the detail of those app servers. 
for instance in WAS you can configure the numbers of retries
https://www.ibm.com/support/knowledgecenter/en/SSEQTP_9.0.5/com.ibm.websphere.base.doc/ae/uejb_timerservice.html

Reply:
does EAP has this configurable as well?

Reply:
nop.